cadence/contracts/services/InterchainTokenService.cdc

```diff
@@ -0,0 +1,324 @@
+
+        self._transmitInterchainTransfer(senderIdentity: senderIdentity, contractName: contractName, contractAddress: contractAddress, destinationChain: destinationChain, destinationAddress: destinationAddress, metadataVersion: 0, metadata: metadata, amount: amount)
+   }
+
+   access(self) fun _takeToken(tokenAddress: Address, tokenName: String, vault: @FungibleToken.Vault){
+        if (self.tokens[tokenAddress] == TokenManagerType.Native){
+            // create nativeToken reference
+            let nativeToken = (&self.nativeTokens[tokenAddress] as &NativeTokens?) ?? panic("could not borrow native token ref")
+            
+            //borrow the coresponding native token vault
+            let tokenVault <- self.account.load<@FungibleToken.Vault>(from: nativeToken.vaultRef)!
+
+            //withdraw from vault and deposit to bridge vault
+            tokenVault.deposit(from: <-vault)
+
+            //save vault back to storage
+            self.account.save(<- tokenVault, to: nativeToken.vaultRef)
+        } else {
+
+            let managedToken = (&self.managedTokens[tokenAddress] as &ManagedTokens?) ?? panic("could not borrow managed token ref")
+
+            let burnCap <- self.account.load<@AxelarFungibleTokenInterface.Burner>(from: managedToken.burnerCapability)!
+
+            burnCap.burnTokens(from: <-vault)
+
+            self.account.save(<- burnCap, to: managedToken.burnerCapability)
+        }
+   }
+
+   access(self) fun _transmitInterchainTransfer(senderIdentity: Capability<&{AxelarGateway.SenderIdentity}>, contractName: String, contractAddress: Address, destinationChain: String, destinationAddress: String, metadataVersion: UInt32, metadata: [UInt8], amount: UFix64){
+        if (metadataVersion > self.LATEST_METADATA_VERSION) {
+            panic("Invalid Metadata Version")
+        }
+
+        let datahash = Crypto.hash(metadata, algorithm: HashAlgorithm.SHA3_256)
+
+        emit InterchainTransfer(
+            tokenAddress: contractAddress,
+            tokenName: contractName,
+            sourceAddress: senderIdentity.address,
+            destinationChain: destinationChain,
+            amount: amount,
+            datahash: datahash
+        )
+
+        //Awaiting ABI.Encode/decode support to parse metadata
+
+        AxelarGateway.callContract(
+            senderIdentity: senderIdentity,
+            destinationChain: destinationChain,
+            destinationContractAddress: destinationAddress,
+            payload: metadata
+        )
+    }
+
+    access(self) fun _processCommand(commandSelector: [UInt8], payload: [UInt8]){
+        //call function based on commandSelector
+    }
+
+    access(self) fun _withdraw(amount: UFix64, tokenAddress: Address, tokenName: String, reciever: &{FungibleToken.Receiver}){
+        if (amount > 0.0){
+            panic("Amount must be greater than zero")
+        }
+        if(!self.tokens.keys.contains(tokenAddress)){
```

Comment:
`contains` will likely run into issues given a sufficient number of mapping entries, but accessing an individual value should scale just fine
```suggestion
        if self.tokens[tokenAddress] == nil {
```